lib/ace/lib/bidiutil.js

```diff
@@ -0,0 +1,556 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL AJAX.ORG B.V. BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * ***** END LICENSE BLOCK ***** */
+
+define(function(require, exports, module) {
+"use strict";
+
+var ArabicAlefBetIntervalsBegine = ['\u0621', '\u0641'];
+var ArabicAlefBetIntervalsEnd = ['\u063A', '\u064a'];
+var dir = 0, hiLevel = 0;
+var lastArabic = false, hasUBAT_AL = false,  hasUBAT_B = false,  hasUBAT_S = false, hasBlockSep = false, hasSegSep = false;
+
+var impTab_LTR = [
+				/*		L,		R,		EN,		AN,		N,		IL,		Cond */
+/* 0 LTR text	*/	[	0,		3,		0,		1,		0,		0,		0	],
+/* 1 LTR+AN		*/	[	0,		3,		0,		1,		2,		2,		0	],
+/* 2 LTR+AN+N	*/	[	0,		3,		0,		0x11,		2,		0,		1	],
+/* 3 RTL text	*/	[	0,		3,		5,		5,		4,		1,		0	],
+/* 4 RTL cont	*/	[	0,		3,		0x15,		0x15,		4,		0,		1	],
+/* 5 RTL+EN/AN	*/	[	0,		3,		5,		5,		4,		2,		0	]
+];
+
+var impTab_RTL = [
+		/*		L,		R,		EN,		AN,		N,		IL,		Cond */
+/* 0 RTL text	*/	[	2,		0,		1,		1,		0,		1,		0	],
+/* 1 RTL+EN/AN	*/	[	2,		0,		1,		1,		0,		2,		0	],
+/* 2 LTR text	*/	[	2,		0,		2,		1,		3,		2,		0	],
+/* 3 LTR+cont	*/	[	2,		0,		2,		0x21,		3,		1,		1	]
+];
+
+var LTR = 0, RTL = 1;
+
+var L = 0; /* left to right				*/
+var R = 1; /* right to left				*/
+var EN = 2; /* European digit				*/
+var AN = 3; /* Arabic-Indic digit			*/
+var ON = 4; /* neutral						*/
+var B = 5; /* block separator				*/
+var S = 6; /* segment separator			*/
+var AL = 7; /* Arabic Letter				*/
+var WS = 8; /* white space					*/
+var CS = 9; /* common digit separator		*/
+var ES = 10; /* European digit separator	*/
+var ET = 11; /* European digit terminator	*/
+var NSM = 12; /* Non Spacing Mark			*/
+var LRE = 13; /* LRE						*/
+var RLE = 14; /* RLE						*/
+var PDF = 15; /* PDF						*/
+var LRO = 16; /* LRO						*/
+var RLO = 17; /* RLO						*/
+var BN = 18; /* Boundary Neutral			*/
+
+var TBBASE = 100, TB00 = 100, TB05 = 101, TB06 = 102, TB07 = 103, TB20 = 104, TBFB = 105, TBFE = 106, TBFF = 107;
+
+var MasterTable = [
+	/************************************************************************************************************************************/
+	/*		0		1		2		3		4		5		6		7		8		9		A	B		C		D		E		F	*/
+	/************************************************************************************************************************************/
+	/*0-*/	TB00,	L	,	L	,	L	,	L	,	TB05,	TB06,	TB07,	R	,	L	,	L	,	L	,	L	,	L	,	L	,	L	,
```

Comment:
i think we should find a way to compress this significantly, otherwise we have to keep it as a separate extension, and not include in ace.js